lib/Conversion/ExportChiselInterface/ExportChiselInterface.cpp

```diff
@@ -0,0 +1,267 @@
+      })
+      .Case<UIntType>([&](UIntType uIntType) {
+        return emitWidthQualifiedType(uIntType, "UInt");
+      })
+      .Case<SIntType>([&](SIntType sIntType) {
+        return emitWidthQualifiedType(sIntType, "SInt");
+      })
+      .Case<AnalogType>([&](AnalogType analogType) {
+        return emitWidthQualifiedType(analogType, "Analog");
+      })
+      .Case<BundleType>([&](BundleType bundleType) {
+        // Emit an anonymous bundle, emitting a `val` for each field
+        return emitTypeWithArguments(
+            "new Bundle ",
+            [&](bool hasEmittedDirection) {
+              os << "{\n";
+              unsigned int nestedIndent = indent + indentIncrement;
+              for (const auto &element : bundleType.getElements()) {
+                os.indent(nestedIndent)
+                    << "val " << element.name.getValue() << " = ";
+                auto elementResult = emitPortType(
+                    location, element.type,
+                    element.isFlip ? direction::flip(direction) : direction, os,
+                    nestedIndent, hasEmittedDirection);
+                if (failed(elementResult))
+                  return failure();
+                os << '\n';
+              }
+              os.indent(indent) << "}";
+              return success();
+            },
+            false);
+      })
+      .Case<FVectorType>([&](FVectorType vectorType) {
+        // Emit a vector type, emitting the type of its element as an argument
+        return emitTypeWithArguments("Vec", [&](bool hasEmittedDirection) {
+          os << vectorType.getNumElements() << ", ";
+          return emitPortType(location, vectorType.getElementType(), direction,
+                              os, indent, hasEmittedDirection);
+        });
+      })
+      .Default([](FIRRTLBaseType) {
+        llvm_unreachable("unknown FIRRTL type");
+        return failure();
+      });
+}
+
+/// Emits an `IO` for the `port`
+static LogicalResult emitPort(const PortInfo &port, llvm::raw_ostream &os) {
+  os.indent(indentIncrement) << "val " << port.getName() << " = IO(";
+  if (failed(emitPortType(port.loc, port.type.cast<FIRRTLBaseType>(),
+                          port.direction, os, indentIncrement)))
+    return failure();
+  os << ")\n";
+
+  return success();
+}
+
+/// Emits an `ExtModule` class with port declarations for `module`
+static LogicalResult emitModule(FModuleLike module, llvm::raw_ostream &os) {
+  os << "class " << module.moduleName() << " extends ExtModule {\n";
+
+  for (const auto &port : module.getPorts()) {
+    if (failed(emitPort(port, os)))
+      return failure();
+  }
+
+  os << "}\n";
+
+  return success();
+}
+
+/// Exports a Chisel interface to the output stream
+static LogicalResult exportChiselInterface(CircuitOp circuit,
+                                           llvm::raw_ostream &os) {
+  // Emit version, package, and import declarations
+  os << circt::getCirctVersionComment() << "package shelf."
```

Comment:
We'll probably want to make the package name configurable at some point but I wasn't sure if it was worth doing yet.

Reply:
Maybe default to the name of the module?

Reply:
Currently it's `shelf.<module-name>`